File: ProgressTracker.cpp
#include "ProgressTracker.hpp"
#include <stdexcept>

// Конструктор: Инициализирует поля
ProgressTracker::ProgressTracker(const std::string& userId, const std::list<std::pair<std::string, double>>& weightLog,
    const std::list<std::string>& sessionHistory, int activeStreak)
    : userId(userId), weightLog(weightLog), sessionHistory(sessionHistory), activeStreak(activeStreak) {}

// Конструктор копирования
ProgressTracker::ProgressTracker(const ProgressTracker& other)
    : userId(other.userId), weightLog(other.weightLog), sessionHistory(other.sessionHistory),
    activeStreak(other.activeStreak) {}

// Деструктор
ProgressTracker::~ProgressTracker() {}

// Метод: Записывает исход сессии
void ProgressTracker::logSessionOutcome(const std::string& outcome) {
    try {
        if (outcome.empty()) {
            throw std::invalid_argument("Результат не может быть пустым");
        }
        this->sessionHistory.push_back(outcome); // Использование this
    }
    catch (const std::exception& e) {
        throw;
    }
}

// Метод: Рассчитывает соблюдение плана(заглушка)
double ProgressTracker::calcCompliance(const std::string& interval) const {
    return 90.0; // Пример
}

// Перегрузка оператора +=
ProgressTracker& ProgressTracker::operator+=(const std::string& session) {
    this->sessionHistory.push_back(session);
    return *this;
}

// Перегрузка оператора []
double ProgressTracker::operator[](const std::string& date) const {
    for (const auto& entry : this->weightLog) {
        if (entry.first == date) {
            return entry.second;
        }
    }
    throw std::out_of_range("Нет данных для этой даты: " + date);
}

File: FitnessAppCsharp/ProgressTracker.cs
using System;
using System.Collections.Generic;

public class ProgressTracker
{
    // Поля класса 
    private string userId;                                          // ID пользователя
    private List<KeyValuePair<string, double>> weightLog;           // История веса: (дата, вес)
    private List<string> sessionHistory;                            // История завершённых сессий
    private int activeStreak;                                       // Текущий стрик (дни подряд)

    // Конструктор
    public ProgressTracker(string userId, List<KeyValuePair<string, double>> weightLog,
        List<string> sessionHistory, int activeStreak)
    {
        this.userId = userId;
        this.weightLog = new List<KeyValuePair<string, double>>(weightLog);     
        this.sessionHistory = new List<string>(sessionHistory);                
        this.activeStreak = activeStreak;
    }

    // Метод: записывает результат завершённой тренировки
    public void LogSessionOutcome(string outcome)
    {
        sessionHistory.Add(outcome);
        activeStreak++; 
        Console.WriteLine($"[Progress] Зафиксирована сессия: {outcome}. Стрик: {activeStreak} дней");
    }

    // Метод: рассчитывает процент соблюдения плана за период (заглушка )
    public double CalcCompliance(string interval)
    {
        return 90.0; // фиксированное значение
    }

    public string GetUserId() => userId;
    public int GetActiveStreak() => activeStreak;
    public List<KeyValuePair<string, double>> GetWeightLog() => weightLog;
    public List<string> GetSessionHistory() => sessionHistory;

 
}